### PR TITLE
feat: optimize local/remote routes flow and store them on Pinecone

### PR DESCRIPTION
**Pull request by tolgadevAI**
### **PR Type**
enhancement

___

### **Description**
- Introduced a new `function_schemas` parameter to the `add` method in the base index class.
- Enhanced the `PineconeRecord` class to include a `function_schema` attribute and updated its `to_dict` method to store this information.
- Modified the `add` method in `PineconeIndex` to handle the new `function_schemas` parameter, ensuring it is processed and stored correctly.
- Updated the `add` method in `semantic_router/layer.py` to pass `function_schemas` to the index, ensuring consistent data handling across layers.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>base.py</strong><dd><code>Add function_schemas parameter to base index add method</code>&nbsp; &nbsp; </dd></summary>
<hr>

semantic_router/index/base.py

- Added `function_schemas` parameter to the `add` method.

</details>

  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/391/files#diff-df154312e842c3302d871397cc51942d0a6e3be19183c0b362499fb8e2fcd127">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pinecone.py</strong><dd><code>Enhance PineconeRecord and add method with function_schemas</code></dd></summary>
<hr>

semantic_router/index/pinecone.py

<li>Added <code>function_schema</code> attribute to <code>PineconeRecord</code>.<br> <li> Updated <code>to_dict</code> method to include <code>sr_function_schemas</code>.<br> <li> Modified <code>add</code> method to handle <code>function_schemas</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/391/files#diff-c57f54339773ad317ee6eafd071395202933f9eeaf59a10c28fc22d06bf564a2">+16/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>layer.py</strong><dd><code>Update layer add method to include function_schemas</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

semantic_router/layer.py

- Updated `add` method to pass `function_schemas` to index.

</details>

  </td>
  <td><a href="https://github.com/aurelio-labs/semantic-router/pull/391/files#diff-e495ddf91e18dd7477eb129fd2c0ab7dfbaf2440534c28b6156a76acdaf51433">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: semantic_router/index/base.py [hunk 1]
```diff
@@ -109,7 +110,11 @@ def delete_index(self):
         raise NotImplementedError("This method should be implemented by subclasses.")
 
     def _sync_index(
-        self, local_route_names: List[str], local_utterances: List[str], dimensions: int
+        self,
+        local_route_names: List[str],
+        local_utterances: List[str],
+        dimensions: int,
+        local_function_schemas: List[str] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/base.py:117]: can we remove `# type: ignore` here and confirm it works as expected?

```suggestion
        local_function_schemas: List[str] | None = None,
```

File: semantic_router/index/local.py [hunk 2]
```diff
@@ -26,6 +27,7 @@ def add(
         embeddings: List[List[float]],
         routes: List[str],
         utterances: List[str],
+        function_schemas: List[Dict[str, Any]] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/local.py:30]: same here, should not use `# type: ignore` — additionally I see `local_function_schemas: List[str]` but here `function_schemas` type is different, why is that?

File: semantic_router/index/base.py [hunk 3]
```diff
@@ -26,6 +26,7 @@ def add(
         embeddings: List[List[float]],
         routes: List[str],
         utterances: List[Any],
+        function_schemas: List[Dict[str, Any]] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/base.py:29]: Should be avoiding use of `# type: ignore` unless required, here we should do:

```suggestion
        function_schemas: List[Dict[str, Any]] | None = None,
```

File: semantic_router/index/pinecone.py [hunk 4]
```diff
@@ -202,7 +207,11 @@ async def _init_async_index(self, force_create: bool = False):
         self.host = index_stats["host"] if index_stats else None
 
     def _sync_index(
-        self, local_route_names: List[str], local_utterances: List[str], dimensions: int
+        self,
+        local_route_names: List[str],
+        local_utterances: List[str],
+        dimensions: int,
+        local_function_schemas: List[str] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/pinecone.py:214]: to remove `type: ignore`

File: semantic_router/index/pinecone.py [hunk 5]
```diff
@@ -305,6 +314,7 @@ def add(
         embeddings: List[List[float]],
         routes: List[str],
         utterances: List[str],
+        function_schemas: List[Dict[str, Any]] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/pinecone.py:317]: remove `type: ignore`

File: semantic_router/index/pinecone.py [hunk 6]
```diff
@@ -313,8 +323,15 @@ def add(
             self.index = self._init_index(force_create=True)
 
         vectors_to_upsert = [
-            PineconeRecord(values=vector, route=route, utterance=utterance).to_dict()
-            for vector, route, utterance in zip(embeddings, routes, utterances)
+            PineconeRecord(
+                values=vector,
+                route=route,
+                utterance=utterance,
+                function_schema=str(function_schema),
```
Comment (jamescalam) [inline semantic_router/index/pinecone.py:330]: in what scenario would `function_schema` not be a string already? If `function_schema` is a `dict` type, should we be using `json.dumps(function_schema)` here?

File: semantic_router/index/postgres.py [hunk 7]
```diff
@@ -255,7 +255,11 @@ def _check_embeddings_dimensions(self) -> bool:
                 raise ValueError("No comment found for the 'vector' column.")
 
     def add(
-        self, embeddings: List[List[float]], routes: List[str], utterances: List[Any]
+        self,
+        embeddings: List[List[float]],
+        routes: List[str],
+        utterances: List[Any],
+        function_schemas: List[Dict[str, Any]] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/postgres.py:262]: remove `type: ignore`

File: semantic_router/index/qdrant.py [hunk 8]
```diff
@@ -165,7 +165,11 @@ def _remove_and_sync(self, routes_to_delete: dict):
             logger.error("Sync remove is not implemented for QdrantIndex.")
 
     def _sync_index(
-        self, local_route_names: List[str], local_utterances: List[str], dimensions: int
+        self,
+        local_route_names: List[str],
+        local_utterances: List[str],
+        dimensions: int,
+        local_function_schemas: List[str] = None,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:172]: remove `type: ignore`

File: semantic_router/layer.py [hunk 9]
```diff
@@ -430,12 +428,17 @@ def add(self, route: Route):
...
 
-        # add routes to the index
         self.index.add(
             embeddings=embeds,
             routes=[route.name] * len(route.utterances),
             utterances=route.utterances,
+            function_schemas=(
+                route.function_schemas * len(route.utterances)
+                if route.function_schemas
+                else [""] * len(route.utterances)  # type: ignore
```
Comment (jamescalam) [inline semantic_router/layer.py:438]: why `type: ignore` here? Can we try to remove

File: semantic_router/layer.py [hunk 10]
```diff
@@ -476,28 +479,47 @@ def _refresh_routes(self):
...
-        embedded_utterances = self.encoder(all_utterances)
-        # create route array
-        # add everything to the index
-        self.index.add(
-            embeddings=embedded_utterances,
-            routes=route_names,
-            utterances=all_utterances,
-        )
+        if routes:
+            for route in routes:
```
Comment (jamescalam) [inline semantic_router/layer.py:483]: why are we iterating through the `routes` here? Didn't seem necessary before?

File: semantic_router/layer.py [hunk 11]
```diff
@@ -430,12 +428,17 @@ def add(self, route: Route):
         if route.score_threshold is None:
             route.score_threshold = self.score_threshold
 
-        # add routes to the index
```
Comment (jamescalam) [inline semantic_router/layer.py:433]: better to keep descriptive comments imo

File: semantic_router/layer.py [hunk 12]
```diff
@@ -216,15 +216,13 @@ def __init__(
         for route in self.routes:
             if route.score_threshold is None:
                 route.score_threshold = self.score_threshold
+
+        if self.routes:
+            self._add_routes(routes=self.routes)
```
Comment (jamescalam) [inline semantic_router/layer.py:221]: this logic seems incorrect, it looks like we try adding routes via `self._add_routes` then a few lines down we may be adding them again with `self._add_and_sync_routes`

@Vits-99 could you also review this section to confirm the functionality?

**Review by jamescalam — CHANGES_REQUESTED** [review]

File: semantic_router/layer.py (hunk 12 above)
Comment (Vits-99) [reply semantic_router/layer.py:221]: Why just not modifying the sync method to accept also function schemas? Or you could associate those function_schemas to route_names and after having synced add that information to the pinecone upsert as metadata associating those function schemas back to the routes to add output from sync method

**Review by jamescalam — APPROVED** [review]

**Pull request merged**